### PR TITLE
`target_sda` will now calculate a consisten target, regardless of the years present in `ald`

### LINKED ISSUE #234: SDA Convergence target differs depending on time-horizon of `ald`
The SDA target is calculated by setting the portfolio average value on a trajectory that converges to the adjusted scenario value, at year 2050. (see here: https://2degreesinvesting.github.io/r2dii.analysis/articles/target-sda.html)

After a conversation with a bank, I noticed some weird behaviour when the target was telling them to increase their emissions, when the target should have decreased. After exploring a bit, it seems that the target calculated is different, depending on the `ald` file. In particular, if the `ald` has data until 2050, then the target is calculated correctly. However, if the ald only has data for a few years into the future, then the target attempts to converge to an earlier year, and if the corporate economy target is greater than the portfolio, it will push it to increase their emissions. 

@2diiKlaus I would appreciate if you could confirm that this doesn't make sense. I think this is a pretty big bug.  

See the reprex below: 

``` r
library(dplyr)
#> 
#> Attaching package: 'dplyr'
#> The following objects are masked from 'package:stats':
#> 
#>     filter, lag
#> The following objects are masked from 'package:base':
#> 
#>     intersect, setdiff, setequal, union
library(ggplot2)
library(r2dii.analysis)

matched <- tibble::tribble(
  ~id_loan, ~loan_size_outstanding, ~loan_size_outstanding_currency, ~loan_size_credit_limit, ~loan_size_credit_limit_currency, ~id_2dii,            ~level, ~score, ~sector,   ~name_ald, ~sector_ald,
  "L1",                      1,                           "EUR",                       2,                            "EUR",    "UP1", "ultimate_parent",      1, "steel", "company a",     "steel"
)

ald <- tibble::tribble(
  ~name_company, ~sector, ~technology, ~year, ~production, ~emission_factor, ~plant_location, ~is_ultimate_owner,
  "company a", "steel",     "steel",  2020,           1,           1.5,            "DE",               TRUE,
  "company a", "steel",     "steel",  2025,           1,           1.5,            "DE",               TRUE,
  "company a", "steel",     "steel",  2030,           1,           1.5,            "DE",               TRUE,
  "company b", "steel",     "steel",  2020,           1,           2.5,            "DE",               TRUE,
  "company b", "steel",     "steel",  2025,           1,           2.5,            "DE",               TRUE,
  "company b", "steel",     "steel",  2030,           1,           2.5,            "DE",               TRUE
)

co2_scenario <- tibble::tribble(
  ~scenario_source, ~scenario, ~sector,  ~region, ~year, ~emission_factor,              ~emission_factor_unit,
  "etp_2017",    "b2ds", "steel", "global",  2020, 2, "tonnes of CO2 per tonne of steel",
  "etp_2017",    "b2ds", "steel", "global",  2025, 1.9, "tonnes of CO2 per tonne of steel",
  "etp_2017",    "b2ds", "steel", "global",  2030, 1.8, "tonnes of CO2 per tonne of steel",
  "etp_2017",    "b2ds", "steel", "global",  2050,      0.25, "tonnes of CO2 per tonne of steel",
)

out1 <- target_sda(matched, ald, co2_scenario)

ggplot(
  data = out1, 
  mapping = aes(
    x = year, 
    y = emission_factor_value, 
    color = emission_factor_metric)
) +
  geom_line() +
  facet_wrap(~ sector)
```

``` r


ald2 <- tibble::tribble(
  ~name_company, ~sector, ~technology, ~year, ~production, ~emission_factor, ~plant_location, ~is_ultimate_owner,
  "company a", "steel",     "steel",  2020,           1,           1.5,            "DE",               TRUE,
  "company a", "steel",     "steel",  2025,           1,           1.5,            "DE",               TRUE,
  "company a", "steel",     "steel",  2030,           1,           1.5,            "DE",               TRUE,
  "company a", "steel",     "steel",  2050,           1,           1.5,            "DE",               TRUE, # add a value for year = 2050
  "company b", "steel",     "steel",  2020,           1,           2.5,            "DE",               TRUE,
  "company b", "steel",     "steel",  2025,           1,           2.5,            "DE",               TRUE,
  "company b", "steel",     "steel",  2030,           1,           2.5,            "DE",               TRUE
)

out2 <- target_sda(matched, ald2, co2_scenario)

ggplot(
  data = out2, 
  mapping = aes(
    x = year, 
    y = emission_factor_value, 
    color = emission_factor_metric)
) +
  geom_line() +
  facet_wrap(~ sector)
```

<sup>Created on 2020-11-30 by the [reprex package](https://reprex.tidyverse.org) (v0.3.0)</sup>

### PR DESCRIPTION
**Pull request by jdhoffa**
Two big changes here: 
- `target_sda()` will now attempt to interpolate between the input values of `scenario`, in case values are missing from the scenario file
- `target_sda()` will now calculate the convergence target to the final year present in `scenario`, regardless of how many years we have data for in `ald`. (Compared to previously, when the convergence target was erroneously calculated to the last year present in `ald`). 

Closes #234

### CODE REVIEW

**Comment by jdhoffa** [comment]
@2diiKlaus compare the following reprex, with that found in #234 

``` r
library(dplyr)
#> 
#> Attaching package: 'dplyr'
#> The following objects are masked from 'package:stats':
#> 
#>     filter, lag
#> The following objects are masked from 'package:base':
#> 
#>     intersect, setdiff, setequal, union
library(ggplot2)
library(r2dii.analysis)

matched <- tibble::tribble(
  ~id_loan, ~loan_size_outstanding, ~loan_size_outstanding_currency, ~loan_size_credit_limit, ~loan_size_credit_limit_currency, ~id_2dii,            ~level, ~score, ~sector,   ~name_ald, ~sector_ald,
  "L1",                      1,                           "EUR",                       2,                            "EUR",    "UP1", "ultimate_parent",      1, "steel", "company a",     "steel"
)

ald <- tibble::tribble(
  ~name_company, ~sector, ~technology, ~year, ~production, ~emission_factor, ~plant_location, ~is_ultimate_owner,
  "company a", "steel",     "steel",  2020,           1,           1.5,            "DE",               TRUE,
  "company a", "steel",     "steel",  2025,           1,           1.5,            "DE",               TRUE,
  "company a", "steel",     "steel",  2030,           1,           1.5,            "DE",               TRUE,
  "company b", "steel",     "steel",  2020,           1,           2.5,            "DE",               TRUE,
  "company b", "steel",     "steel",  2025,           1,           2.5,            "DE",               TRUE,
  "company b", "steel",     "steel",  2030,           1,           2.5,            "DE",               TRUE
)

co2_scenario <- tibble::tribble(
  ~scenario_source, ~scenario, ~sector,  ~region, ~year, ~emission_factor,              ~emission_factor_unit,
  "etp_2017",    "b2ds", "steel", "global",  2020, 2, "tonnes of CO2 per tonne of steel",
  "etp_2017",    "b2ds", "steel", "global",  2025, 1.9, "tonnes of CO2 per tonne of steel",
  "etp_2017",    "b2ds", "steel", "global",  2030, 1.8, "tonnes of CO2 per tonne of steel",
  "etp_2017",    "b2ds", "steel", "global",  2050,      0.25, "tonnes of CO2 per tonne of steel",
)

out1 <- target_sda(matched, ald, co2_scenario)

ggplot(
  data = out1, 
  mapping = aes(
    x = year, 
    y = emission_factor_value, 
    color = emission_factor_metric)
) +
  geom_line() +
  facet_wrap(~ sector)
```



``` r

ald2 <- tibble::tribble(
  ~name_company, ~sector, ~technology, ~year, ~production, ~emission_factor, ~plant_location, ~is_ultimate_owner,
  "company a", "steel",     "steel",  2020,           1,           1.5,            "DE",               TRUE,
  "company a", "steel",     "steel",  2025,           1,           1.5,            "DE",               TRUE,
  "company a", "steel",     "steel",  2030,           1,           1.5,            "DE",               TRUE,
  "company a", "steel",     "steel",  2050,           1,           1.5,            "DE",               TRUE, # add a value for year = 2050
  "company b", "steel",     "steel",  2020,           1,           2.5,            "DE",               TRUE,
  "company b", "steel",     "steel",  2025,           1,           2.5,            "DE",               TRUE,
  "company b", "steel",     "steel",  2030,           1,           2.5,            "DE",               TRUE
)

out2 <- target_sda(matched, ald2, co2_scenario)

ggplot(
  data = out2, 
  mapping = aes(
    x = year, 
    y = emission_factor_value, 
    color = emission_factor_metric)
) +
  geom_line() +
  facet_wrap(~ sector)
```



<sup>Created on 2020-12-02 by the [reprex package](https://reprex.tidyverse.org) (v0.3.0)</sup>

File: DESCRIPTION
Comment (maurolepore) [inline DESCRIPTION:45]: I see a new dependency from which we use only one function. It it's simple, I would try to re-implement it (maybe see `zoo:::na.approx.default`). If not so easy, I think it's ok as this package has few if any non-base dependencies, and its maintainers seem very active.

File: R/target_sda.R
Comment (maurolepore) [inline R/target_sda.R:304]: Why this change? Just curious if it relates to some issue other than interpolation.

**Review by maurolepore — APPROVED** [review]
Looks like you know what you are doing :) and I see it's well covered by tests. 
Assuming it's not trivial to avoid the new dependency, it LGTM.

File: R/target_sda.R
Comment (jdhoffa) [reply R/target_sda.R:304]: this is the crucial change for:
> target_sda() will now calculate the convergence target to the final year present in scenario, regardless of how many years we have data for in ald. (Compared to previously, when the convergence target was erroneously calculated to the last year present in all).

`inner_join` would only keep track of the scenario values if there was a corresponding year in `ald`, but we need to know scenario values for any year.

Comment (jdhoffa) [reply R/target_sda.R:304]: `full_join` allows us to retain all info from both datasets (which is desirable)

File: DESCRIPTION
Comment (jdhoffa) [reply DESCRIPTION:45]: it doesn't look super trivial to me and seems to depend on other functions available in `zoo`

Comment (jdhoffa) [reply DESCRIPTION:45]: but i can continue to look! just exploring #239 right now

**Comment by maurolepore** [comment]
Thanks! I prefer if you are the once how merges -- as you are better informed to write a meaningful squash-commit message.

**Comment by jdhoffa** [comment]
Ok, happy to do so

**Pull request merged**
